### PR TITLE
guard against case 2 drag from underneath case 3 1 user 2 tabs

### PR DESCRIPTION
**Pull request by mattkrick**
TEST

CASE 2
- start a retro, add 3+ reflections, go to group phase
- Browser 1 drops card A on card B
- [x] Browser 2 picks up card B before card A completes its drop on it  (you have 1 second, go fast!) and drops it on card C
- the board is as expected: card A is in a group by itself, card B & C are in a group
Note: I added a 1 second delay on the server to make testing this easier. We'll need to remove that before we merge this!

CASE 3
- open 2 tabs for the same user
- in tab 1, drop card A on card B
- quickly try to pick up card A & ungroup it. keep trying until you have success
- [x] tab 2 does not get an "interception" toast
- [ ] tab 2 can drag card A  

AutoGroup
- ungroup all the cards
- click autogroup
- [x] the groups autogroup themselves & a refresh is not needed

## Notes

OMG. This is one million times better. Incredible work!

I pounded on this hard this morning. I still see a case every once in a great while where a frame must drop and one of the clients gets a hanging card. It is fixed on a refresh. Of course, we know that's possible so nothing to do about it now.

A couple of other things of note:

### Cards not reflowing on window resize event?

See gif below:

Is this is a regression? Not that the cards do not reflow when I resize, and must horizontal scroll to access them:

[gif](http://g.recordit.co/ArdDVSA2OE.gif)

^ looks like I can no longer embed big gifs :(

Shall we fix this now?

### Auto-grouping one-time failure

I ran 3 retros, each with different card content. On one run, I hit autogroup and about half of my cards disappeared. A refresh did not fix it. Since then, I've been unable to reproduce it.

Very odd. Something to keep an eye on. If we make a subsequent revision of this PR, I'll try to reproduce it.

### CODE REVIEW

File: src/server/graphql/mutations/endDraggingReflection.js [hunk 1]
```diff
@@ -84,6 +85,7 @@ export default {
       dragId
     }
 
+    await sleep(1000)
```
Comment (mattkrick) [inline src/server/graphql/mutations/endDraggingReflection.js]: REMOVE ME AFTER TESTING

File: src/universal/components/PhaseItemMasonry.js [hunk 2]
```diff
@@ -146,6 +146,11 @@ class PhaseItemMasonry extends React.Component<Props> {
       const {x, y} = bestEl.getBoundingClientRect()
       setClosingTransform(atmosphere, itemId, {x, y})
     }
+    if (atmosphere.startDragQueue && atmosphere.startDragQueue.length) {
+      // reply the startDrag event that fired before we received the end drag event
+      const queuedStart = atmosphere.startDragQueue.shift()
+      queuedStart()
```
Comment (mattkrick) [inline src/universal/components/PhaseItemMasonry.js]: this is just a robust version of handling case 1. i still prefer the locking until the ack comes back because without locking there are still other edge cases.

**Comment by jordanh** [comment]
# Test Cycle 1

## CASE 2
- start a retro, add 3+ reflections, go to group phase
- Browser 1 drops card A on card B
- [ ] Browser 2 picks up card B before card A completes its drop on it  (you have 1 second, go fast!) and drops it on card C
- the board is as expected: card A is in a group by itself, card B & C are in a group
Note: I added a 1 second delay on the server to make testing this easier. We'll need to remove that before we merge this!

## CASE 3
- open 2 tabs for the same user
- in tab 1, drop card A on card B
- quickly try to pick up card A & ungroup it. keep trying until you have success
- [ ] tab 2 does not get an "interception" toast
- [ ] tab 2 can drag card A  

## AutoGroup
- ungroup all the cards
- click autogroup
- [ ] the groups autogroup themselves & a refresh is not needed

File: package.json [hunk 3]
```diff
@@ -208,7 +208,7 @@
     "tinycolor2": "^1.4.1",
     "tlds": "^1.192.0",
     "unicode-substring": "^1.0.0",
-    "uws": "^10.148.0"
+    "uws": "10.148.0"
```
Comment (jordanh) [inline package.json]: Pegged to this version as there is a squabble going on. This package version is immutable

**Comment by mattkrick** [comment]
ready for another go! @jordanh

**Comment by jordanh** [comment]
# Test Cycle 2

## CASE 2
- start a retro, add 3+ reflections, go to group phase
- Browser 1 drops card A on card B
- [ ] Browser 2 picks up card B before card A completes its drop on it  (you have 1 second, go fast!) and drops it on card C
- the board is as expected: card A is in a group by itself, card B & C are in a group
Note: I added a 1 second delay on the server to make testing this easier. We'll need to remove that before we merge this!

## CASE 3
- open 2 tabs for the same user
- in tab 1, drop card A on card B
- quickly try to pick up card A & ungroup it. keep trying until you have success
- [ ] tab 2 does not get an "interception" toast
- [ ] tab 2 can drag card A  

## AutoGroup
- ungroup all the cards
- click autogroup
- [ ] the groups autogroup themselves & a refresh is not needed

**Comment by mattkrick** [comment]
EXTRA TEST:
- [ ] the facilitator moves forward
- [ ] someone else groups a card
- [ ] the facilitator's cards are correct

File: src/universal/components/NewMeeting.js [hunk 4]
```diff
@@ -288,6 +289,18 @@ class NewMeeting extends Component<Props> {
...
 
+  onSidebarTransitionEnd = (e) => {
+    const {
+      atmosphere: {eventEmitter},
+      viewer: {
+        team: {isMeetingSidebarCollapsed}
+      }
+    } = this.props
+    if (e.target === this.sidebarRef.current && e.propertyName === 'transform') {
+      eventEmitter.emit('meetingSidebarCollapsed', isMeetingSidebarCollapsed)
```
Comment (mattkrick) [inline src/universal/components/NewMeeting.js]: the alternative to this is
- put a cDU on the masonry root
- see if the prop changed, and if it did, reinit the grid after a setTimeout

this is cleaner because if we change the duration of the sidebar toggle, the group grid doesn't break.

**Comment by jordanh** [comment]
Is this ready for my 👀  Mr. @mattkrick ?

**Comment by mattkrick** [comment]
yep!

**Comment by jordanh** [comment]
# TEST

- [x] the facilitator moves forward
- [x] someone else groups a card
- [x] the facilitator's cards are correct 

## NOTES

### New exception encountered when re-entering meeting:

This is non-critical to fix in this PR:

https://github.com/ParabolInc/action/issues/2306

**Comment by jordanh** [comment]
Woo hoo! LGTM! Merging this.
